cmake/SetupFormaline.cmake

```diff
@@ -0,0 +1,81 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+# When running formaline when SpECTRE was not built from a Git repository
+# we need a list of all the files and directories in the root of the source
+# directory that are tracked by Git. I.e.
+#   git ls-tree --full-tree --name-only HEAD
+set(SPECTRE_FORMALINE_LOCATIONS
+  ".clang-format;cmake;CMakeLists.txt;containers;docs;external;"
+  ".github;.gitignore;LICENSE.txt;README.md;src;.style.yapf;"
+  "support;tests;tools;.travis;.travis.yml")
+
+find_package(Git)
+
+if(EXISTS ${CMAKE_SOURCE_DIR}/.git AND Git_FOUND)
+  execute_process(
+    COMMAND ${GIT_EXECUTABLE} ls-tree --full-tree --name-only HEAD
+    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+    OUTPUT_VARIABLE FORMALINE_GIT_FILES_TRACKED
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+    )
+  string(REPLACE "\n" ";" FORMALINE_GIT_FILES_TRACKED
+    "${FORMALINE_GIT_FILES_TRACKED}")
+
+  # Check all elements in SPECTRE_FORMALINE_LOCATIONS
+  # are in FORMALINE_GIT_FILES_TRACKED. We don't just compare that
+  # the lists are equal because we don't have a guarantee that the list
+  # stored in CMake and the list output by the call to git will be
+  # in the same order.
+  foreach(FILE ${SPECTRE_FORMALINE_LOCATIONS})
+    list(FIND FORMALINE_GIT_FILES_TRACKED ${FILE} FOUND_FILE)
+    if(${FOUND_FILE} EQUAL -1)
+      message(FATAL_ERROR
+        "Couldn't find the file or directory \"${FILE}\" in the "
+        "source directory. You need to remove the entry \"${FILE}\" "
+        "from the CMake variable SPECTRE_FORMALINE_LOCATIONS in "
+        "cmake/SetupFormaline.cmake")
+    endif()
+  endforeach()
+
+  # Check all elements in FORMALINE_GIT_FILES_TRACKED
+  # are in SPECTRE_FORMALINE_LOCATIONS. We don't just compare that
+  # the lists are equal because we don't have a guarantee that the list
+  # stored in CMake and the list output by the call to git will be
+  # in the same order.
+  foreach(FILE ${FORMALINE_GIT_FILES_TRACKED})
+    list(FIND SPECTRE_FORMALINE_LOCATIONS ${FILE} FOUND_FILE)
+    if(${FOUND_FILE} EQUAL -1)
+      message(FATAL_ERROR
+        "Couldn't find the file or directory \"${FILE}\" in the "
+        "CMake variable SPECTRE_FORMALINE_LOCATIONS in "
+        "cmake/SetupFormaline.cmake. You need to add \"${FILE}\" "
+        "to the CMake variable SPECTRE_FORMALINE_LOCATIONS in "
+        "cmake/SetupFormaline.cmake")
+    endif()
+  endforeach()
```

Comment:
Could you instead check that `list(SORT A)` equals `list(SORT B)` ? This should avoid any ordering-related inequalities, without the hassle of checking every element of A is in B, then checking every element of B is in A...

Reply:
Yep, I think that _should_ work (CMake :sweat_smile: ). However, I still think I prefer the current for a different reason, which is that currently you get printed out exactly what the missing or extra entry is. Doing that with CMake is as easily done and would amount to about the same amount of code. I haven't found what would effectively be a "list difference", though it could be done with the "list(FILTER..." I think. Thoughts?

If we decide to keep as-is, I'll update the comment to reflect the decision.

Reply:
Hmm, from a quick look at the docs, I agree that doing this using CMake list functions may end up needing just as much code (but CMake.. so maybe I'm missing something). I'd be satisfied for you to simply clarify the reasoning in the comment 🙂